rfcs/0001-task-chunking.md

```diff
@@ -0,0 +1,440 @@
+while still allowing the application to run tasks in parallel.
+
+![chunks and stickiness](.images/0001-task-chunking-chunks-with-stickiness.svg)
+
+The above diagrams show tasks that always take the exact same time. This is rarely
+true. There will be a drop in parallelism when tasks are finishing at the end of
+a chunk, without new tasks to start. Here's how that looks.
+
+![chunks and stickiness with random](.images/0001-task-chunking-chunks-with-stickiness-random-task-times.svg)
+
+Users can select larger chunks to amortize this drop in parallelism, but there are
+ways to address this remaining inefficency while still allowing a scheduler to handle
+tasks individually and provide responsive load balancing. One possibility is
+to introduce a mechanism that pipelines tasks from the scheduler to the step script.
+Such a mechanism would add more complexity than the present chunking proposal,
+and is left for future work.
+
+## Specification
+
+> Changes to [the template schema](https://github.com/OpenJobDescription/openjd-specifications/wiki/2023-09-Template-Schemas).
+
+> A modification to [`<TaskParameterDefinition>`](https://github.com/OpenJobDescription/openjd-specifications/wiki/2023-09-Template-Schemas#341-taskparameterdefinition)
+
+```diff
+  <TaskParameterDefinition> ::= <IntTaskParameterDefinition> | <FloatTaskParameterDefinition> |
+-                               <StringTaskParameterDefinition> | <PathTaskParameterDefinition>
++                               <StringTaskParameterDefinition> | <PathTaskParameterDefinition> |
++                               <ChunkIntTaskParameterDefinition>
+```
+
+> A new section `<ChunkIntTaskParameterDefinition>` after
+> [section 3.4.1.4. `<PathTaskParameterDefinition>`](https://github.com/OpenJobDescription/openjd-specifications/wiki/2023-09-Template-Schemas#3414-pathtaskparameterdefinition).
+
+### 3.4.1.5. `<ChunkIntTaskParameterDefinition>`
+
+```diff
++  An integer valued task parameter, processed as chunks instead of as individual elements.
++  A `<ChunkIntTaskParameterDefinition>` is the object:
++
++ ```yaml
++   name: <Identifier>
++   type: "CHUNK[INT]"
++   range: <IntRangeList> | <IntRangeExpr>
++   chunks:
++     defaultTaskCount: <integer> | <intstring> # @fmtstring
++     targetRuntimeSeconds: <integer> | <intstring> # @optional @fmtstring
++     rangeConstraint: "CONTIGUOUS" | "NONCONTIGUOUS"
++ ```
++
++ See section `<IntTaskParameterDefinition>` for the definitions of `<IntRangeExpr>` and `<intstring>`.
++
++ 1. *name* — The name of the parameter.
++ 2. *type* — The literal "CHUNK[INT]", defining this parameter as integer valued and processed
++     as chunks.
++ 3. *range* — The list of values that the parameter takes on to define Tasks of the Step.
++ 4. *chunks* — Specifies how to form sets of values into chunks.
++     1. *defaultTaskCount* — How many tasks to combine into a single chunk by default.
```

Comment:
Why are we using the `default` prefix? Is it because this value can be overridden at runtime based on `targetRuntimeSeconds` or by the remaining frames in a range? If so, is this the same as `maxTaskCount`?

Reply:
Thanks for reviewing the proposal! It's called default to indicate the scheduler should use this value if it doesn't have a better value to use. It's not the same as `maxTaskCount` because the scheduler can choose larger values, like when chunks are finishing quicker than in `targetRuntimeSeconds`.